cluster-autoscaler/simulator/cluster.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+
+	kube_api "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	cmd "k8s.io/kubernetes/pkg/kubectl/cmd"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// Nodes with utilization below this are considered unused and mayb be subject to scale down.
+	unusedThreshold = float64(0.5)
+)
+
+// FindNodeToRemove finds a node that can be removed.
+func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube_client.Client) (*kube_api.Node, error) {
+
+	nodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+
+	for _, node := range nodes {
+		nodeInfo, found := nodeNameToNodeInfo[node.Name]
+		if !found {
+			glog.Errorf("Node info for %s not found", node.Name)
+			continue
+		}
+		reservation, err := calculateReservation(node, nodeInfo, kube_api.ResourceCPU)
+		if err != nil {
+			glog.Warningf("Failed to calculate reservation for %s: %v", node.Name, err)
+		}
+		glog.V(4).Infof("Node %s - reservation %f", node.Name, reservation)
+
+		if reservation > unusedThreshold {
+			glog.Infof("Node %s is not suitable for removal - reservation to big (%f)", node.Name, reservation)
+			continue
+		}
+		//Lets try to remove this one.
+		glog.V(2).Infof("Considering %s for removal", node.Name)
+
+		podsToRemoveList, _, _, err := cmd.GetPodsForDeletionOnNodeDrain(client, node.Name,
+			kube_api.Codecs.UniversalDecoder(), false, true)
+
+		if err != nil {
+			glog.V(1).Infof("Node %s cannot be removed: %v", node.Name, err)
+			continue
+		}
+
+		tempNodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+		delete(tempNodeNameToNodeInfo, node.Name)
+		ptrPodsToRemove := make([]*kube_api.Pod, 0, len(podsToRemoveList))
+		for i := range podsToRemoveList {
+			ptrPodsToRemove = append(ptrPodsToRemove, &podsToRemoveList[i])
+		}
+
+		findProblems := findPlaceFor(ptrPodsToRemove, nodes, tempNodeNameToNodeInfo)
+		if findProblems == nil {
+			return node, nil
+		} else {
+			glog.Infof("Node %s is not suitable for removal %v", node.Name, err)
+		}
+	}
+	return nil, nil
+}
+
+func calculateReservation(node *kube_api.Node, nodeInfo *schedulercache.NodeInfo, resourceName kube_api.ResourceName) (float64, error) {
```

Comment:
I'm pretty sure this logic is implemented somewhere in the kubelet, so it'd great to reuse it (not in this PR).